cmd/yggd/worker.go

```diff
@@ -65,27 +278,52 @@ func watchWorkerDir(dir string, env []string, died chan int) {
 		log.Debugf("received inotify event %v", e.Event())
 		switch e.Event() {
 		case notify.InCloseWrite, notify.InMovedTo:
-			if strings.HasSuffix(e.Path(), "worker") {
-				if ExcludeWorkers[filepath.Base(e.Path())] {
+			name := filepath.Base(e.Path())
+			if validWorker(name) {
+				var config *workerConfig
+
+				configFile := filepath.Join(workerConfigDir, name+".toml")
+				if fileExists(configFile) {
+					var err error
+
+					config, err = loadWorkerConfig(configFile)
+					if err != nil {
+						log.Errorf("cannot read worker config file: %v", err)
+						continue
+					}
+				} else {
+					config = &workerConfig{}
+				}
+				config.program = filepath.Join(workerExecDir, name)
+
+				if err := startWorker(*config, nil, func(pid int) {
+					died <- pid
+				}); err != nil {
+					log.Errorf("cannot start worker %v: %v", config.program, err)
 					continue
 				}
-				log.Tracef("new worker detected: %v", e.Path())
-				go startProcess(e.Path(), env, 0, died)
 			}
 		case notify.InDelete, notify.InMovedFrom:
 			workerName := filepath.Base(e.Path())
-			pidFilePath := filepath.Join(
-				yggdrasil.LocalstateDir,
-				"run",
-				yggdrasil.LongName,
-				"workers",
-				workerName+".pid",
-			)
-
-			if err := killWorker(pidFilePath); err != nil {
-				log.Errorf("cannot kill worker: %v", err)
+			if err := stopWorker(workerName); err != nil {
+				log.Errorf("cannot stop worker: %v", err)
 				continue
 			}
 		}
 	}
 }
+
+func validEnvVar(val string) bool {
+	for _, pattern := range []string{"PATH=.*", "YGG_.*=.*"} {
+		r := regexp.MustCompile(pattern)
+		if r.Match([]byte(val)) {
```

Comment:
I would print some warning log that given environment variable was filtered out.

Reply:
Good idea.